src/data/sum/order.lean

```diff
@@ -0,0 +1,228 @@
+⟨by { rintro _ (⟨a, _, h⟩ | ⟨a, _, h⟩); exact irrefl _ h }⟩
+
+instance [is_trans α r] [is_trans β s] : is_trans (α ⊕ β) (lift_rel r s) :=
+⟨by { rintro _ _ _ (⟨a, b, hab⟩ | ⟨a, b, hab⟩) (⟨_, c, hbc⟩ | ⟨_, c, hbc⟩),
+  exacts [lift_rel.inl (trans hab hbc), lift_rel.inr (trans hab hbc)] }⟩
+
+instance [is_antisymm α r] [is_antisymm β s] : is_antisymm (α ⊕ β) (lift_rel r s) :=
+⟨by { rintro _ _ (⟨a, b, hab⟩ | ⟨a, b, hab⟩) (⟨_, _, hba⟩ | ⟨_, _, hba⟩); rw antisymm hab hba }⟩
+
+end lift_rel
+
+section lex
+variables (r : α → α → Prop) (s : β → β → Prop)
+
+instance [is_refl α r] [is_refl β s] : is_refl (α ⊕ β) (lex r s) :=
+⟨by { rintro (a | a), exacts [lex.inl (refl _), lex.inr (refl _)] }⟩
+
+instance [is_irrefl α r] [is_irrefl β s] : is_irrefl (α ⊕ β) (lex r s) :=
+⟨by { rintro _ (⟨a, _, h⟩ | ⟨a, _, h⟩); exact irrefl _ h }⟩
+
+instance [is_trans α r] [is_trans β s] : is_trans (α ⊕ β) (lex r s) :=
+⟨by { rintro _ _ _ (⟨a, b, hab⟩ | ⟨a, b, hab⟩) (⟨_, c, hbc⟩ | ⟨_, c, hbc⟩),
+  exacts [lex.inl (trans hab hbc), lex.sep _ _, lex.inr (trans hab hbc), lex.sep _ _] }⟩
+
+instance [is_antisymm α r] [is_antisymm β s] : is_antisymm (α ⊕ β) (lex r s) :=
+⟨by { rintro _ _ (⟨a, b, hab⟩ | ⟨a, b, hab⟩) (⟨_, _, hba⟩ | ⟨_, _, hba⟩); rw antisymm hab hba }⟩
+
+instance [is_total α r] [is_total β s] : is_total (α ⊕ β) (lex r s) :=
+⟨λ a b, match a, b with
+  | inl a, inl b := (total_of r a b).imp lex.inl lex.inl
+  | inl a, inr b := or.inl (lex.sep _ _)
+  | inr a, inl b := or.inr (lex.sep _ _)
+  | inr a, inr b := (total_of s a b).imp lex.inr lex.inr
+end⟩
+
+instance [is_trichotomous α r] [is_trichotomous β s] : is_trichotomous (α ⊕ β) (lex r s) :=
+⟨λ a b, match a, b with
+  | inl a, inl b := (trichotomous_of r a b).imp3 lex.inl (congr_arg _) lex.inl
+  | inl a, inr b := or.inl (lex.sep _ _)
+  | inr a, inl b := or.inr (or.inr $ lex.sep _ _)
+  | inr a, inr b := (trichotomous_of s a b).imp3 lex.inr (congr_arg _) lex.inr
+end⟩
+
+instance [is_well_order α r] [is_well_order β s] : is_well_order (α ⊕ β) (sum.lex r s) :=
+{ wf := sum.lex_wf is_well_order.wf is_well_order.wf }
+
+end lex
+
+/-! ### Disjoint sum of two orders -/
+
+section disjoint
+
+instance [has_le α] [has_le β] : has_le (α ⊕ β) := ⟨lift_rel (≤) (≤)⟩
+instance [has_lt α] [has_lt β] : has_lt (α ⊕ β) := ⟨lift_rel (<) (<)⟩
+
+lemma le_def [has_le α] [has_le β] {a b : α ⊕ β} : a ≤ b ↔ lift_rel (≤) (≤) a b := iff.rfl
+lemma lt_def [has_lt α] [has_lt β] {a b : α ⊕ β} : a < b ↔ lift_rel (<) (<) a b := iff.rfl
+
+lemma inl_le_inl_iff [has_le α] [has_le β] {a b : α} : (inl a : α ⊕ β) ≤ inl b ↔ a ≤ b :=
+lift_rel_inl_inl
+
+lemma inr_le_inr_iff [has_le α] [has_le β] {a b : β} : (inr a : α ⊕ β) ≤ inr b ↔ a ≤ b :=
+lift_rel_inr_inr
+
+lemma inl_lt_inl_iff [has_lt α] [has_lt β] {a b : α} : (inl a : α ⊕ β) < inl b ↔ a < b :=
+lift_rel_inl_inl
+
+lemma inr_lt_inr_iff [has_lt α] [has_lt β] {a b : β} : (inr a : α ⊕ β) < inr b ↔ a < b :=
+lift_rel_inr_inr
```

Comment:
`simp`? It might also be nice to have a `inr_strict_mono` statement of this lemma.